text/3512-closure-move-bindings.md

```diff
@@ -0,0 +1,284 @@
+without any parentheses in between.
+
+## _ImplicitReference_ closures
+
+When the parentheses for _MoveBindings_ is present, or when the `move` keyword is absent,
+the closure expression is of the _ImplicitReference_ type, where
+all local variables in the closure construction scope not shadowed by any _MoveBinding_
+are implicitly captured into the closure by shared or mutable reference on demand,
+preferring shared reference if possible.
+
+Each _MoveBinding_ declares binding(s) in its left-side pattern,
+assigned with the value of the right-side expression evaluated during closure construction,
+thus referencing any relevant local variables if necessary.
+
+If the left-side pattern is omitted (_UnnamedMoveBinding_),
+the expression must be either a single-segment (identifier) `PathExpression`
+or a _MethodCallExpression_,
+the receiver expression of which must be a single identifier variable,
+and the argument list must not reference any local variables.
+The left-side pattern is then automatically inferred to be a simple _IdentifierPattern_
+using the identifier/receiver as the new binding.
+
+### Mutable bindings
+
+If a captured binding mutated inside the closure is declared in a _NamedMoveBinding_,
+the `IdentifierPattern` that declares the binding must have the `mut` keyword.
+
+If it is declared in an _UnnamedMoveBinding_,
+the `mut` keyword must be added in front of the expression;
+since the declared binding is always the first token in the expression,
+the `mut` token is always immediately followed by the mutable binding,
+thus yielding consistent readability.
+
+If it is implicitly captured from the parent scope
+instead of declared in a _MoveBinding_,
+the local variable declaration must be declared `mut` too.
+
+## _ImplicitMove_ closures
+
+When the `move` keyword is present but _MoveBindings_ is absent (with its parentheses absent as well),
+the closure expression is of the _ImplicitMove_ type, where
+all local variables in the closure construction scope
+are implicitly moved or copied into the closure on demand.
+
+Note that `move` with an empty pair of parentheses is allowed and follows the former rule;
+in other words, `move() |...| {...}` and `|...| {...}` are semantically equivalent.
+This allows macros to emit repeating groups of `_MoveBinding_ ","` inside a pair of parentheses
+and achieve correct semantics when there are zero repeating groups.
+
+If a moved binding is mutated inside the closure,
+its declaration in the parent scope must be declared `mut` too.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Due to backwards compatibility, this RFC proposes a new syntax
+that is an extension of capture-by-move
+but actually looks more similar to capture-by-reference,
+thus confusing new users.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Capture-by-reference is the default behavior for implicit captures for two reasons:
```

Comment:
I'd like to suggest the opposite, here, actually: for `move(..)` (or `move()` if my suggestion for `..` is disregrejected) to mean the same as `move`.

  - big consistency gain: `move(..) ||` being closer to `move ||` than to `||`;
  - make by-reference captures be the odd one that stands out. I agree this may be subjective, but bear with me:

    There are two kind of closures:
      - immediately-called closures, such as with `Option::map()` or `panic::catch_unwind()`;
      - long-lived closures (`: 'long`), such as the `spawn()` family.
    The former, 99% of the time, accomodates to fully implicit `|| {}` capture rules, and the latter, kind of accomodate to `move ||` closures, except for when specific "right-before `move` adjustments" are needed, which is _very often_ just a `.clone()` (and sometimes an `Arc::downgrade()`).

    This very RFC, by providing explicit captures, is thus mostly helping the latter case: the one with `: 'long`-lived requirements. In such a scenario, capturing by reference is rather non-desired, or something niche stemming from a long-lived borrowable (_e.g._, a variable with outlives the `'env` of a `scope::spawn()`)). It thus does seem rather legitimate to favor by-`move` captures for the `..`-implicit ones, much like `move ||` does.

  - lastly, although I agree this may not be that important to others, I feel like it would be the most helpful syntax w.r.t. teachability: having the word `move` not be neutered by some explicit `&` sigil (_e.g._, in `move(&x) || x + 42)`) mean a by-ref capture kind of denaturates/dilutes the meaning of the word `move`

Reply:
I think, rather than adding `move(..)` (which could just be written `move`) or `move()` (which could just be omitted), we should have a warn-by-default lint for an empty capture list (with a suggestion that removes it and a note that mentions using `move` if you want to capture everything), and add it to the list of lints suppressed in macros (since macros might want to generate a capture list).